scripts/update_institution_sso_email_domain.py

```diff
@@ -0,0 +1,117 @@
+    FAILED = 2  # The email failed to be added since it belongs to another account
+    ERRORED = 3  # The email failed to be added due to unexpected exceptions
+
+
+def main():
+    """Update the emails of users from the affiliated institution (when eligible). Requires three arguments:
+    `institution_id`, `src_domain` and `dst_domain`. The domains do not go through email validation or spam check
+    since the script can only be run by trusted and privileged admins.
+    """
+
+    args = parser.parse_args()
+    institution_id = args.id
+    src_domain = args.src
+    dst_domain = args.dst
+
+    # Verify the institution
+    institution = Institution.load(institution_id)
+    if not institution:
+        message = 'Error: invalid institution ID [{}]'.format(institution_id)
+        logger.error(message)
+        sentry.log_message(message)
+        return
+
+    # Find all users affiliated with the given institution
+    affiliated_users = OSFUser.objects.filter(affiliated_institutions___id=institution_id)
+
+    # Update email domain for each user
+    succeeded_users = []
+    skipped_users = []
+    failed_users = []
+    errored_users = []
+    for user in affiliated_users:
+        result = update_email_domain(user, src_domain, dst_domain)
+        if result == UpdateResult.SUCCEEDED:
+            succeeded_users.append(user._id)
+        if result == UpdateResult.SKIPPED:
+            skipped_users.append(user._id)
+        if result == UpdateResult.FAILED:
+            failed_users.append(user._id)
+        if result == UpdateResult.ERRORED:
+            errored_users.append(user._id)
+
+    # Output update results to console
+    logger.info(f'{UpdateResult.SUCCEEDED.name} ({len(succeeded_users)}/{len(affiliated_users)}) = {succeeded_users}')
+    logger.warning(f'{UpdateResult.SKIPPED.name} ({len(skipped_users)}/{len(affiliated_users)}) = {skipped_users}')
+    logger.error(f'{UpdateResult.FAILED.name} ({len(failed_users)}/{len(affiliated_users)}) = {failed_users}')
+    logger.error(f'{UpdateResult.ERRORED.name} ({len(errored_users)}/{len(affiliated_users)}) = {errored_users}')
+
+
+def update_email_domain(user, src_domain, dst_domain):
+    """For a given user, if it has an email `example@<src_domain>`, attempt to add `example@<dst_domain>` to this
+    account. The action will be skipped if `example@<dst_domain>` already exists on this user or no email is found
+    under `@<src_domain>`. The action will fail if `example@<dst_domain>` already exists on another account. Other
+    unexpected exceptions will be considered as error.
+    """
+
+    # Find the eligible email to add
+    email_to_add = None
+    for email in user.emails.all():
+        email_parts = email.address.split('@')
+        name_part = email_parts[0].lower()
+        domain_part = email_parts[1].lower()
+        if domain_part == src_domain:
+            email_to_add = f'{name_part}@{dst_domain}'
+            break
```

Comment:
Implausible, but what if the user has multiple emails at the `src_domain` in question already? Should this update all of them?

Reply:
I agree. The better solution is to add every eligible email instead of the first one found and I will make the change.

If the institution only migrate one of their emails (e.g. the primary email), then we will end up with extra invalid emails. However, by adding all of them, at least will will have the correct one added.

I am drafting an email to them with other questions and will include this note. In the future, we will have another script to make the new email primary. We may not be able to decide which one to use if they have multiple.

Reply:
I need to re-do the output part since `1` -> `N`  changed how successes/failures/errors are stored and presented.